### DISCUSSION TITLE
Setting prefetch=0 on request.dynamic doesn't work, but prefetch=1 does?

### DISCUSSION THREAD

**Discussion by ScuttleSE** [savonet/liquidsoap#2113] [Q&A]
I have a source that looks like this

`gbsfm = request.dynamic(prefetch=1, id="gbsfm", get_request)`

setting prefetch to 1 makes the source work, but not the way I want. when starting LS it makes two requests to get_request, for the current song, and next song. At the end of current song, next song is played and one new request is done.

I want the initial request to only ask get_request once

I thought that was what the prefetch=0 would do, but it seems to just... disable the entire source

**Comment by toots** [comment]
Hi,

This is the default, yes. If you need more involved implementations, you can define your own. Each request dynamic source has a `add` method that you can use to push a new request whenever you see fit. You can also see the whole queue with `queue` and set the queue at once with `set_queue`.

This way, you can create a source, add an initial track and queue the next track when it gets near the end. Something like this:
```ruby
s = request.dynamic(prefetch=0, ...)

if not s.add(some_initial_request) then
  log("Error while queueing initial request!")
end

def on_end(_, _) =
  new_request = ...
  if not s.add(new_request) then
    log("Error while queueing next request!")
  end
end

s = source.on_end(s, on_end)
```

Be aware, though, that doing things like that can be tricky, in particular if it takes too long to prepare the next track or if your source is behind a switch/fallback and may not get played for a while etc.

The general philosophy of our implementation designs is to choose a default implementation that is simple and fits most people's need but leave it open to implement advanced uses like this.. 🙂

**Reply by ScuttleSE** [reply]
In the first line of your example you set prefetch=0

When I do the same in my script, it just ignores that source

I havde a short script that gets URIs to files

```
def get_request() =
  result = process.read.lines("python3 /home/liq/getsong.py")
  log("next song: #{result}")
  request.create(list.hd(default="", result))
end

```

I then have a source attached to that

`gbsfm = request.dynamic(prefetch=1, id="gbsfm", get_request)`

Is that not the way to do it?

That's how I have been running it for over a year with the 1.4 versions. Now that I run the exact same script, it seems to have reverted to prefetch one song, with no way to turn it off?

So, in short; can I use the script I have been using and somehow disable prefetching, or do I have to rewrite it?

**Reply by toots** [reply]
Yes, unfortunately the algorithm to prefetch songs in `request.dynamic` was rewritten and simplified with the `2.0` release. While it worked for you, it was causing a lot of issues with other users.

**Reply by ScuttleSE** [reply]
so Yes I have to rewrite it?

**Reply by ScuttleSE** [reply]
Okay... so in your example, line 3, you have "some_initial_request"

Can I use my old script here? Or what is that?

You then have new_request in the on_end definition. Is that another script? or can I use my old one there again?

**Reply by ScuttleSE** [reply]
Also, in your example you're defining on_end, isn't on_end an already existing function?

**Reply by toots** [reply]
I think with your script it'll look like this:
```ruby
def get_request() =
  result = process.read.lines("python3 /home/liq/getsong.py")
  log("next song: #{result}")
  request.create(list.hd(default="", result))
end

s = request.dynamic(prefetch=0,  id="gbsfm", get_request)

if not s.add(get_request()) then
  log("Error while queueing initial request!")
end

def on_end(_, _) =
  new_request = get_request()
  if not s.add(new_request) then
    log("Error while queueing next request!")
  end
end

s = source.on_end(s, on_end)
```

**Reply by rrolla** [reply]
trying to use your script but got type error:

Liquidsoap v2.0.3

```ruby
music = request.dynamic(id="music", prefetch=0, retry_delay=20., timeout=20., dynamic_request)

if not music.add(dynamic_request()) then
  log("Error while queueing initial request!")
end

def on_end(_, _) =
  new_request = dynamic_request()
  if not s.add(new_request) then
    log("Error while queueing next request!")
  end
end

music = source.on_end(music, on_end)
```

Error:
```ruby
rolla-liquidsoap | if not music.add(dynamic_request()) then
rolla-liquidsoap | 
rolla-liquidsoap | Error 5: this value has type
rolla-liquidsoap |   _? (inferred at /home/liquidsoap/config/extras/api-fetch.liq, line 41, char 12-18)
rolla-liquidsoap | but it should be a subtype of
rolla-liquidsoap |   request
```

**Reply by rrolla** [reply]
Sorry @toots most likely you did not notice this post ^

**Reply by toots** [reply]
What's the code at `/home/liquidsoap/config/extras/api-fetch.liq`, line `41`?

**Reply by rrolla** [reply]
@toots here is part of api-fetch.liq
```ruby
38        if request.resolve(r) then
39            r
40        else
41            null()
42       end
```

**Reply by toots** [reply]
you need to handle the case where your API does not return a request:
```ruby
def add() =
  r = dynamic_request()
  if null.defined(r) then
    music.add(null.get(r))
  else
    false
  end
end

if not add() then
  log("Error while queueing initial request!")
end

def on_end(_, _) =
  if not add() then
    log("Error while queueing next request!")
  end
end
```

**Reply by rrolla** [reply]
thanks will try

**Reply by rrolla** [reply]
Yes, this works after renaming your function `add` to something else, huge thanks @toots

**Comment by ScuttleSE** [comment]
Here's the beginning of my LS-file

```
set("harbor.bind_addr","0.0.0.0")
set("server.telnet.bind_addr", "0.0.0.0")
set("server.telnet",true)
set("log.file",false)
set("log.stdout",true)
set("log.level",3)

def get_request() =
  result = process.read.lines("python3 /home/liq/getsong.py")
  log("next song: #{result}")
  request.create(list.hd(default="", result))
end

def auth_gbsfm(params)
  user = params.user
  password = params.password
  ret = process.read.lines("python3 /home/liq/streampw.py #{password}")
  ret = list.hd(default="",ret)
  if ret == "true" then
    true
  else
    false
  end
end

#Live source

live = input.harbor("gbsfm_live",port=8889,auth=auth_gbsfm)

#Website playlist source
#gbsfm = request.dynamic(prefetch=1, id="gbsfm", get_request)

gbsfm = request.dynamic(prefetch=0,  id="gbsfm", get_request)

if not gbsfm.add(get_request()) then
  log("Error while queueing initial request!")
end

def on_end(_, _) =
  new_request = get_request()
  if not gbsfm.add(new_request) then
    log("Error while queueing next request!")
  end
end

gbsfm = source.on_end(on_end, gbsfm)

#Fallback
beees = single("/home/liq/bees.mp3")

#Make sure mono files are being played
gbsfm = audio_to_stereo(gbsfm)

#Put them all together
gbsfm = fallback(track_sensitive=false, [live, gbsfm, beees])

```

Running it gives me an error on the on_end part

```
At ices.liq, line 47, char 22-28:
Error 5: this value has type
  (_, _) -> _ (inferred at ices.liq, line 40 char 0 - line 45 char 3)
but it should be a subtype of
  source(_)

```

**Reply by toots** [reply]
`on_end` and `s` arguments need to be swapped, I updated my script examples above.

**Reply by rrolla** [reply]
I got a different error: https://github.com/savonet/liquidsoap/discussions/2113#discussioncomment-2352242 click on the link and  reload page it will open discussion
